bfvmm/include/hve/arch/intel_x64/delegator/cpuid.h

```diff
@@ -0,0 +1,157 @@
+//
+// Copyright (C) 2019 Assured Information Security, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef BFVMM_DELEGATOR_CPUID_INTEL_X64_H
+#define BFVMM_DELEGATOR_CPUID_INTEL_X64_H
+
+#include <stdint.h>
+#include <bfdelegate.h>
+#include <bfgsl.h>
+#include <bftypes.h>
+#include <unordered_map>
+#include <list>
+
+namespace bfvmm::intel_x64::cpuid
+{
+
+/// CPUID leaf data type
+///
+///
+using leaf_t = uint64_t;
+
+/// Info
+///
+/// This struct is created by cpuid::delegator::handle before being
+/// passed to each registered handler.
+///
+struct info_t {
+
+    /// RAX (in/out)
+    ///
+    uint64_t rax;
+
+    /// RBX (in/out)
+    ///
+    uint64_t rbx;
+
+    /// RCX (in/out)
+    ///
+    uint64_t rcx;
+
+    /// RDX (in/out)
+    ///
+    uint64_t rdx;
+
+    /// Ignore write (out)
+    ///
+    /// If true, do not update the guest's register state with the four
+    /// register values above. Set this to true if you do not want the guest
+    /// rax, rbx, rcx, or rdx to be written to after your handler completes.
+    ///
+    /// default: false
+    ///
+    bool ignore_write;
+
+    /// Ignore advance (out)
+    ///
+    /// If true, do not advance the guest's instruction pointer.
+    /// Set this to true if your handler returns true and has already
+    /// advanced the guest's instruction pointer.
+    ///
+    /// default: false
+    ///
+    bool ignore_advance;
+};
+
+using delegate_t = delegate<bool(vcpu_t, info_t &)>;
+#define handler(function) delegate_t::create<function>()
```

Comment:
if these are generic, they should be moved to their own header. If these are not, the name is to generic and will collide with other headers that define the same thing.

Is there a way to do this using templates and auto?

Reply:
These may or may not be generic, I don't think we can tell until all the handers are moved over.

I agree there is probably a better way to do this, I'd like to push off solving that problem until a future PR when we start deprecating old handlers.